### PR TITLE
chore: add s11 to discovery with it's added logic

### PR DESCRIPTION
**Pull request by YogevBokobza**
## Description

>In this code we are adding s11 to the discovery part + it's added logic to support that + refactoring get function to support more than 1 runner

**Related issue (if any):** fixes #issue_number_goes_here

## Checklist

- [x] I have followed this repository's contributing guidelines.
- [x] I will adhere to the project's code of conduct.

## Additional information

> Anything else?

### CODE REVIEW

**Comment by TheSpringBoy** [comment]
Hi, is it possible to add support to the light switches as well? Since you already dealt with the lights here..

**Comment by YogevBokobza** [comment]
> Hi, is it possible to add support to the light switches as well? Since you already dealt with the lights here..

It is not related to this PR..
I rather not to add more parts and I am trying to have the minimum code every time so we can make a progress faster..

File: src/aioswitcher/api/messages.py [hunk 1]
```diff
@@ -130,16 +131,22 @@ def get_thermostat_remote_id(self) -> str:
         remote_hex = unhexlify(self._hex_response)
         return remote_hex[84:92].decode().rstrip("\x00")
 
-    def get_shutter_position(self) -> int:
-        """Return the current shutter position."""
-        hex_pos = self._hex_response[152:154].decode()
+    def get_shutter_position(self, index: int) -> int:
+        """Return the current position of the shutter 0 <= pos <= 100."""
+        index -= 1
```
Comment (thecode) [inline src/aioswitcher/api/messages.py]: I see that the `index -= 1` is repeating in multiple places here, wouldn't it be better to use zero based indexing? after all this is something internal and we programmers preferer counting from zero.

Comment (YogevBokobza) [reply src/aioswitcher/api/messages.py]: Done.

Comment (thecode) [reply src/aioswitcher/api/messages.py]: > Done.

Where exactly is it done?

Comment (YogevBokobza) [reply src/aioswitcher/api/messages.py]: https://github.com/TomerFi/aioswitcher/pull/767/files#diff-76fda1abe206ffc78578e59e694e42cf998f9158ec5427d72e3f42d8db31c7a4R226
Here I can see the change..

Comment (YogevBokobza) [reply src/aioswitcher/api/messages.py]: https://github.com/TomerFi/aioswitcher/pull/767/commits/630a921d800c2fdd81306cf031b9f4e5397b104c

Comment (YogevBokobza) [reply src/aioswitcher/api/messages.py]: Oh this is something else.. let me see

Comment (YogevBokobza) [reply src/aioswitcher/api/messages.py]: In this case, I can't change this part..
This is part of the logic..
If the index is 1 then it should be 0 if this is 2 then it should be 1 and so on..
This is how this works..

Comment (thecode) [reply src/aioswitcher/api/messages.py]: Everywhere in this PR you reduce 1 from the index (about 6 occasion), than the only place you can get the index is a method you implanted which can return 1, 3 or 0 (which was returning -1 that I also don't understand why it is now changed to 0 and what is it used for).
Why can't this method return 0 or 2 so you don't have to reduce it by one later everywhere?

> This is how this works..

That is not how PR review works, there is a method that return -1 (changed to 0 without any explanation why), 3 or 1 without any documentation what it means, you have no plan to explain why you wrote it like that, I don't mind we resolve this conversation and I let someone else review it.

Comment (YogevBokobza) [reply src/aioswitcher/api/messages.py]: I will try to explain..
The changes here: https://github.com/TomerFi/aioswitcher/pull/767/commits/630a921d800c2fdd81306cf031b9f4e5397b104c
I changed based on your request as here it does not matter if this is -1 or 0 because if the conditions are not applied it does not matter what is returning.

The other index -= 1 are parts of this logic to return the correct packet and it should remain..

Comment (thecode) [reply src/aioswitcher/api/messages.py]: I think we are having a broken discussion. I didn't ask to return 0 instead of -1 from `get_shutter_index`, if this condition can't be met you should either return `None` or raise.

I asked that we use zero based index instead of having to do `index -= 1` on all of the methods here. I don't see this breaking any method as instead of passing 3 to a method and let it subtract 1 to get 2, you can just pass 2.
I fully understand from the code here how the packets are generated and I don't understand your comment:
> The other index -= 1 are parts of this logic to return the correct packet and it should remain..

So please explain it better if I am missing something.

Comment (thecode) [inline src/aioswitcher/api/messages.py]:
```suggestion
```

**Review by thecode — CHANGES_REQUESTED** [review]
Please use zero based index. Fix methods that are used to get the index to count from zero so the code will have the same result.

File: src/aioswitcher/api/messages.py (hunk 1 above)
Comment (YogevBokobza) [reply src/aioswitcher/api/messages.py]: Honestly I don't agree with this change.
You have in your house zero and one shutters?
No you have one and two.
I want go and get shutter number one and shutter number 2.
Using 0 and 1 make no sense here..

Using 0 as a starting point makes sense for indexing that outside of human logic mainly for array, but here you are talking about the shutter number you want to control.
If you buy one Switcher Runner you can control shutter with id 1 which is one shutter you have.

Comment (thecode) [reply src/aioswitcher/api/messages.py]: Why does it have to be related to what the user see? User can still see the first shutter as 1, it would probably mean that we will have one or two places that we have to add one when we display it to the user, but it doesn't make sense that all low level APIs will just reduce 1 from the input, they are not exposed to the user

Comment (TomerFi) [reply src/aioswitcher/api/messages.py]: I agree with @thecode on this. Starting an indexed list from 1 instead of 0 might be in line with the user's perspective, but this is error-prone. A contributor might add some logic and look for the 0 index member. 

Also, changing the value of an argument from the function's signature is a bad practice unless the function is meant to be used recursively. But this is irrelevant if you switch to zero-based indexing and dump the `index -= 1` statement.

File: src/aioswitcher/device/tools.py [hunk 2]
```diff
@@ -211,3 +213,24 @@ def validate_token(username: str, token: str) -> bool:
     else:
         logger.debug("request failed with status code: %s", response.status_code)
     return is_token_valid
+
+
+def get_shutter_index(device_type: DeviceType, device_num: int) -> int:
```
Comment (TomerFi) [inline src/aioswitcher/device/tools.py]: This is very weird and requires some explanation in the code docs.

File: src/aioswitcher/device/tools.py [hunk 3]
```diff
@@ -211,3 +213,24 @@ def validate_token(username: str, token: str) -> bool:
...
+        if device_num == 1:
+            return 1
+    if device_type == DeviceType.RUNNER_S11:
+        if device_num == 1:
+            return 3
+    return 0
+
+
+def get_light_index(device_type: DeviceType, device_num: int) -> int:
+    """Return the currect light index (based of device type)."""
```
Comment (TomerFi) [inline src/aioswitcher/device/tools.py]: Please explain this.

File: src/aioswitcher/device/__init__.py [hunk 4]
```diff
@@ -121,6 +121,31 @@ def value(self) -> str:
         return self._value  # type: ignore
 
 
+@final
+class LightState(Enum):
```
Comment (TomerFi) [inline src/aioswitcher/device/__init__.py]: Is this identical to DeviceState?

**Review by TomerFi — CHANGES_REQUESTED** [review]

File: src/aioswitcher/api/messages.py (hunk 1 above)
Comment (YogevBokobza) [reply src/aioswitcher/api/messages.py]: Change has been done..

Comment (YogevBokobza) [reply src/aioswitcher/api/messages.py]: Change has been done + added testes

File: src/aioswitcher/device/tools.py (hunk 2 above)
Comment (YogevBokobza) [reply src/aioswitcher/device/tools.py]: Added more docs to the functions description.

File: src/aioswitcher/device/tools.py (hunk 3 above)
Comment (YogevBokobza) [reply src/aioswitcher/device/tools.py]: I will explain.
get_shutter_index is used to get the correct position to look for in the packet for getting the shutter's direction/position.
get_light_index is used to get the correct position to look for in the packet for getting the light status if it's on or off.

The main idea of the position is like this:

- RUNNER/RUNNER MINI:
The light index does not exist
The shutter index is 0
- Runner S11:
The first light index is 0
The second light index is 1
The shutter index is 2
- Runner S12:
The light index is 0
The first shutter index is 1
The second shutter index is 2

So basically, if the device has lights, it comes before the shutter.

I could remove get_shutter_index and get_light_index altogether and put these hardcoded values inside the bridge.py file.
I just figured it cleaner to look on this part like this.

File: src/aioswitcher/device/__init__.py (hunk 4 above)
Comment (YogevBokobza) [reply src/aioswitcher/device/__init__.py]: Yes and no..
DeviceState is meant to show the device's state..
But let's say Runner S11 which Runner is the device but it has 2 different modules in it which is 2 lights.
I think it makes more sense to have LightState for each of them because they are not devices inside the device..

File: src/aioswitcher/api/messages.py [hunk 5]
```diff
@@ -130,16 +131,20 @@ def get_thermostat_remote_id(self) -> str:
         remote_hex = unhexlify(self._hex_response)
         return remote_hex[84:92].decode().rstrip("\x00")
 
-    def get_shutter_position(self) -> int:
-        """Return the current shutter position."""
-        hex_pos = self._hex_response[152:154].decode()
+    def get_shutter_position(self, index: int) -> int:
+        """Return the current position of the shutter 0 <= pos <= 100."""
+        start_index = 152 + (index * 32 if index else 0)
```
Comment (thecode) [inline src/aioswitcher/api/messages.py]: The `if index` has no effect
```suggestion
        start_index = 152 + (index * 32)
```

File: src/aioswitcher/api/messages.py [hunk 6]
```diff
@@ -130,16 +131,20 @@ def get_thermostat_remote_id(self) -> str:
...
         return int(hex_pos, 16)
 
-    def get_shutter_direction(self) -> ShutterDirection:
-        """Return the current shutter direction."""
-        hex_dir = self._hex_response[156:160].decode()
-        directions = dict(map(lambda s: (s.value, s), ShutterDirection))
-        return directions[hex_dir]
+    def get_shutter_direction(self, index: int) -> ShutterDirection:
+        """Return the current direction of the shutter (UP/DOWN/STOP)."""
+        start_index = 156 + (index * 32 if index else 0)
```
Comment (thecode) [inline src/aioswitcher/api/messages.py]: The `if index` has no effect
```suggestion
        start_index = 156 + (index * 32)
```

File: src/aioswitcher/bridge.py [hunk 7]
```diff
@@ -380,19 +411,33 @@ def get_device_type(self) -> DeviceType:
...
         return devices[hex_model]
 
-    # Switcher Runner and Runner Mini methods
+    # Switcher Runners methods
 
-    def get_shutter_position(self) -> int:
+    def get_shutter_position(self, index: int) -> int:
         """Return the current position of the shutter 0 <= pos <= 100."""
-        hex_pos = hexlify(self.message[135:137]).decode()
+        start_index = 135 if index == 0 else 135 + (index * 16)
```
Comment (thecode) [inline src/aioswitcher/bridge.py]: Same, `if index` is not needed

File: src/aioswitcher/bridge.py [hunk 8]
```diff
@@ -380,19 +411,33 @@ def get_device_type(self) -> DeviceType:
...
+        end_index = start_index + 2
+        hex_pos = hexlify(self.message[start_index:end_index]).decode()
         return int(hex_pos[2:4]) + int(hex_pos[0:2], 16)
 
-    def get_shutter_direction(self) -> ShutterDirection:
+    def get_shutter_direction(self, index: int) -> ShutterDirection:
         """Return the current direction of the shutter (UP/DOWN/STOP)."""
-        hex_direction = hexlify(self.message[137:139]).decode()
+        start_index = 137 if index == 0 else 137 + (index * 16)
```
Comment (thecode) [inline src/aioswitcher/bridge.py]: Same, `if index` is not needed

File: src/aioswitcher/bridge.py [hunk 9]
```diff
@@ -380,19 +411,33 @@ def get_device_type(self) -> DeviceType:
...
+        end_index = start_index + 2
+        hex_direction = hexlify(self.message[start_index:end_index]).decode()
         directions = dict(map(lambda d: (d.value, d), ShutterDirection))
         return directions[hex_direction]
 
+    def get_light_state(self, index: int) -> LightState:
+        """Extract the light state from the broadcast message."""
+        start_index = 135 if index == 0 else 135 + (index * 16)
```
Comment (thecode) [inline src/aioswitcher/bridge.py]: Same, `if index` is not needed

**Review by thecode — CHANGES_REQUESTED** [review]

File: src/aioswitcher/bridge.py (hunk 9 above)
Comment (YogevBokobza) [reply src/aioswitcher/bridge.py]: Fixed..

File: src/aioswitcher/bridge.py (hunk 8 above)
Comment (YogevBokobza) [reply src/aioswitcher/bridge.py]: Fixed..

File: src/aioswitcher/bridge.py (hunk 7 above)
Comment (YogevBokobza) [reply src/aioswitcher/bridge.py]: Fixed..

File: src/aioswitcher/api/messages.py [hunk 10]
```diff
@@ -130,16 +131,20 @@ def get_thermostat_remote_id(self) -> str:
...
-        hex_dir = self._hex_response[156:160].decode()
-        directions = dict(map(lambda s: (s.value, s), ShutterDirection))
-        return directions[hex_dir]
+    def get_shutter_direction(self, index: int) -> ShutterDirection:
+        """Return the current direction of the shutter (UP/DOWN/STOP)."""
+        start_index = 156 + (index * 32)
+        end_index = start_index + 4
+        hex_direction = self._hex_response[start_index:end_index].decode()
+        directions = dict(map(lambda d: (d.value, d), ShutterDirection))
+        return directions[hex_direction]
```
Comment (thecode) [inline src/aioswitcher/api/messages.py]: A general comment not a request for changes, the following 3 lines renaming variables without any need, please try to avoid it in feature PRs and separate such changes if they are needed to code quality PRs. 
In this PR 2 lines would not even show on the review if the name `hex_dir` was kept as it is.

File: src/aioswitcher/device/tools.py [hunk 11]
```diff
@@ -211,3 +213,32 @@ def validate_token(username: str, token: str) -> bool:
...
     return is_token_valid
+
+
+def get_shutter_index(device_type: DeviceType, device_num: int) -> int:
+    """Return the correct shutter index.
+
+    Used in retriving the shutter position/direction from the packet
+    (based of device type and device number).
+    """
+    if device_type == DeviceType.RUNNER or device_type == DeviceType.RUNNER_MINI:
```
Comment (thecode) [inline src/aioswitcher/device/tools.py]: I can only assume this is a preparation for future devices not `S11` but for now it doesn't make sense to check `device_num` in each conditon.
```python
if device_num != 0:
    raise ValueError("Invalid device number")

if device_type in (DeviceType.RUNNER, DeviceType.RUNNER_MINI):
   return 0
elif device_type == DeviceType.RUNNER_S11:
  return 2

raise ValueError("only shutters are allowed")
```

File: src/aioswitcher/api/messages.py (hunk 10 above)
Comment (YogevBokobza) [reply src/aioswitcher/api/messages.py]: Should I revert to hex_dir or leave it like this?

File: src/aioswitcher/device/tools.py (hunk 2 above)
Comment (thecode) [reply src/aioswitcher/device/tools.py]: I think even after the update it is not clear.
Please add a small part that explain the device configuration. something like:
S11 has two lights circuits & one shutter circuits.
Lights circuits are numbered 0 & 1, shutter circuit is 0.

I also think the name `device_num` is confusing as we use device in multiple places here. I don't know what would be the best option but maybe `circuit_number` or `sub_device_number`
I usually try to avoid short names in external params, I prefer it will be clear what it does vs a shorter code.

File: src/aioswitcher/device/__init__.py [hunk 12]
```diff
@@ -321,6 +346,23 @@ class SwitcherShutterBase(ABC):
...
+        position: the current position of the shutter (integer percentage).
+        direction: the current direction of the shutter.
+        light: the current light state.
+        light2: the current second light state.
+    """
+
+    position: int
+    direction: ShutterDirection
+    light: LightState
+    light2: LightState
```
Comment (thecode) [inline src/aioswitcher/device/__init__.py]: Maybe it would be better to make this an array?

File: src/aioswitcher/api/messages.py (hunk 10 above)
Comment (thecode) [reply src/aioswitcher/api/messages.py]: I think it would be better to revert to `hex_dir` just to make it simpler to understand the changes in this message.

File: src/aioswitcher/device/__init__.py (hunk 12 above)
Comment (YogevBokobza) [reply src/aioswitcher/device/__init__.py]: It is possible but I think it makes things more complicated and not that much readable if you consider the maximum number of lights possible in a device is 3 using the product: Switcher Light SL03

File: src/aioswitcher/device/tools.py (hunk 2 above)
Comment (YogevBokobza) [reply src/aioswitcher/device/tools.py]: Regarding:
`
S11 has two lights circuits & one shutter circuits.
Lights circuits are numbered 0 & 1, shutter circuit is 0.
`
Where is the best place to write this?

Comment (YogevBokobza) [reply src/aioswitcher/device/tools.py]: @thecode I put this information next to those functions.. I am not sure where to put it if not here..
Also changed to circuit_number

File: src/aioswitcher/device/tools.py (hunk 11 above)
Comment (YogevBokobza) [reply src/aioswitcher/device/tools.py]: Changed to the code suggested.

File: src/aioswitcher/device/__init__.py [hunk 13]
```diff
@@ -376,3 +418,15 @@ def __post_init__(self) -> None:
...
         return super().__post_init__()
+
+
+@final
+@dataclass
+class SwitcherSingleShutterDualLight(SwitcherSingleShutterDualLightBase, SwitcherBase):
+    """Implementation of the Switcher Shutter with dual light device."""
+
+    def __post_init__(self) -> None:
+        """Post initialization validate device type category as SINGLE_SHUTTER_DUAL_LIGHT."""  # noqa: E501
```
Comment (thecode) [inline src/aioswitcher/device/__init__.py]: Please break into two lines instead of adding #noqa

File: src/aioswitcher/device/tools.py (hunk 2 above)
Comment (thecode) [reply src/aioswitcher/device/tools.py]: From my point it looks clear now

File: src/aioswitcher/device/__init__.py (hunk 13 above)
Comment (YogevBokobza) [reply src/aioswitcher/device/__init__.py]: I tried to change to
"""Post initialization validate device type category
        as SINGLE_SHUTTER_DUAL_LIGHT.
        """

but got this lint errors:
src/aioswitcher\device\__init__.py:429:1: D205 1 blank line required between summary line and description
src/aioswitcher\device\__init__.py:429:1: D400 First line should end with a period

Comment (thecode) [reply src/aioswitcher/device/__init__.py]: That is correct, when documenting a function you should have a short description on the first line, than a black line and after you can multiple lines, something like this should work:
```
"""Post initialization

Validate device type category as SINGLE_SHUTTER_DUAL_LIGHT."""
```

File: src/aioswitcher/device/__init__.py (hunk 12 above)
Comment (thecode) [reply src/aioswitcher/device/__init__.py]: So when you add the 3rd light it will be `light3`? wouldn't it be better to have an array, than when you add additional light to the array it will just expand by one?

I leave @TomerFi to suggest what he thinks would be better here

Comment (YogevBokobza) [reply src/aioswitcher/device/__init__.py]: I think an array is unneeded here and can make even more confusions, but let's leave it for @TomerFi

File: src/aioswitcher/device/__init__.py (hunk 13 above)
Comment (YogevBokobza) [reply src/aioswitcher/device/__init__.py]: src/aioswitcher\device\__init__.py:429:1: D209 Multi-line docstring closing quotes should be on a separate line
src/aioswitcher\device\__init__.py:429:1: D400 First line should end with a period

Still issue..

Comment (thecode) [reply src/aioswitcher/device/__init__.py]: Well I just gave an example, yes first line should end with a period and you need to move the 3 """ to a new line.
It seems strange you come back here with this, wouldn't it be simple just to fix it the errors? they are clear.
To be honest I never face these errors since I am using linters which reformat the text as I write it.

Comment (YogevBokobza) [reply src/aioswitcher/device/__init__.py]: Done..

File: src/aioswitcher/device/__init__.py (hunk 4 above)
Comment (TomerFi) [reply src/aioswitcher/device/__init__.py]: IDK... It feels weird two identical classes just for their names.
@thecode What do you think?

File: src/aioswitcher/device/__init__.py (hunk 12 above)
Comment (TomerFi) [reply src/aioswitcher/device/__init__.py]: What is the chance we'll need to add another light slot? If we add more, I agree with @thecode: Let's switch to an array. But if it's a long shot, we can stick to vars, maybe change the name to "lightSlot1" and "lightSlot2" or something like that because light+light2 feels like a patch.

Another option might be to have a dict with a string key ("light_slot1", "light_slot2") and a State instance as a value.

Comment (thecode) [reply src/aioswitcher/device/__init__.py]: We will need another slot in the future see https://github.com/TomerFi/aioswitcher/pull/767#discussion_r1615241270, Switcher Light SL03 has 3 lights.

File: src/aioswitcher/device/__init__.py (hunk 4 above)
Comment (thecode) [reply src/aioswitcher/device/__init__.py]: I think we can start without it and if we face a problem we can add it later.

File: src/aioswitcher/device/__init__.py (hunk 12 above)
Comment (TomerFi) [reply src/aioswitcher/device/__init__.py]: Gotcha.
Let's do an array. it feels right.

File: src/aioswitcher/device/__init__.py (hunk 4 above)
Comment (YogevBokobza) [reply src/aioswitcher/device/__init__.py]: Removed and changed to DeviceState..

File: src/aioswitcher/device/__init__.py (hunk 12 above)
Comment (YogevBokobza) [reply src/aioswitcher/device/__init__.py]: I changed that to an array.

**Review by thecode — DISMISSED** [review]

File: src/aioswitcher/device/tools.py [hunk 14]
```diff
@@ -211,3 +213,43 @@ def validate_token(username: str, token: str) -> bool:
...
+    """Return the correct light index.
+
+    Used in retriving the light on/off status from the packet
+    (based of device type and circuit number).
+    """
+    if device_type == DeviceType.RUNNER_S11:
+        if circuit_number == 0:
+            return 0
+        elif circuit_number == 1:
+            return 1
```
Comment (TomerFi) [inline src/aioswitcher/device/tools.py]:
```python
if circuit_number in [0, 1]:
    return circuit_number
```

**Review by TomerFi — CHANGES_REQUESTED** [review]

File: src/aioswitcher/device/tools.py (hunk 14 above)
Comment (YogevBokobza) [reply src/aioswitcher/device/tools.py]: Done..

**Review by thecode — APPROVED** [review]

**Review by TomerFi — APPROVED** [review]

**Comment by TomerFi** [comment]
@YogevBokobza @thecode - Thank you very much!
I'm merging this.

Also, we need to rebase runner-support-fixed from the main.
